### PR TITLE
packages/background

### PR DESCRIPTION
**Pull request by marxeille**
## Summary by CodeRabbit

- **New Features**
  - Introduced a new message type and handler for simulating Tron signing requests.
  - Enhanced `KeyRingService` to support Tron transaction simulation.
  - Added utility functions for feature detection in blockchain environments.
  - Added support for a new currency and improved integration with multiple blockchain technologies.

- **Bug Fixes**
  - Improved error logging and handling in various services.

- **Refactor**
  - Updated variable declarations for better scoping and clarity.
  - Streamlined response structures and control flows in multiple services.
  - Refactored method parameters and return types in `KeyRingService`.

- **Chores**
  - Minor formatting and organization improvements for better code readability.

### CODE REVIEW

**Review by haunv3 — APPROVED** [review]

File: packages/background/src/keyring/keyring.ts [hunk 1]
```diff
@@ -298,6 +302,14 @@ export class KeyRing {
       throw new Error("Key Store is empty");
     }
 
+    // // Need to check network type by chain id instead of coin type
```
Comment (toandq2009) [inline packages/background/src/keyring/keyring.ts]: Need remove comment note

File: packages/background/src/keyring/keyring.ts [hunk 2]
```diff
@@ -1507,18 +1536,21 @@ export class KeyRing {
     chainId,
     defaultCoinType,
   }: {
-    typedMessage: V extends "V1" ? TypedDataV1 : TypedMessage<T>;
+    // typedMessage: V extends "V1" ? TypedDataV1 : TypedMessage<T>;
```
Comment (toandq2009) [inline packages/background/src/keyring/keyring.ts]: need remove comment this

File: packages/background/src/keyring/keyring.ts [hunk 3]
```diff
@@ -1507,18 +1536,21 @@ export class KeyRing {
...
+    typedMessage: string;
     version: V;
     chainId: string;
     defaultCoinType: number;
-  }): ECDSASignature {
+    // }): Promise<ECDSASignature> {
```
Comment (toandq2009) [inline packages/background/src/keyring/keyring.ts]: need remove comment note this

File: packages/background/src/keyring/keyring.ts [hunk 4]
```diff
@@ -1507,18 +1536,21 @@ export class KeyRing {
...
+  }): Promise<string> {
     try {
       this.validateVersion(version);
       if (!typedMessage) {
         throw new Error("Missing data parameter");
       }
 
-      const coinType = this.computeKeyStoreCoinType(chainId, defaultCoinType);
+      // const coinType = this.computeKeyStoreCoinType(chainId, defaultCoinType);
```
Comment (toandq2009) [inline packages/background/src/keyring/keyring.ts]: need remove comment note

File: packages/background/src/keyring/messages.ts [hunk 5]
```diff
@@ -139,6 +155,30 @@ export class ShowKeyRingMsg extends Message<string> {
...
+  public static type() {
+    return "simulate-sign-tron";
+  }
+
+  constructor(public readonly msg: any) {
+    super();
+  }
+
+  validateBasic(): void {
+    // if (!this.msg) {
```
Comment (toandq2009) [inline packages/background/src/keyring/messages.ts:168]: need remove comment note

File: packages/background/src/keyring/service.ts [hunk 6]
```diff
@@ -634,12 +660,34 @@ export class KeyRingService {
...
+        version: SignTypedDataVersion.V4,
+        chainId,
+        defaultCoinType: 60,
+      });
+
+      return rawTxHex;
+    } catch (e) {
+      console.log("e", e.message);
+    } finally {
+      // this.interactionService.dispatchEvent(APP_PORT, "request-sign-end", {});
```
Comment (toandq2009) [inline packages/background/src/keyring/service.ts:682]: need remove comment note

File: packages/background/src/keyring/service.ts [hunk 7]
```diff
@@ -809,21 +817,21 @@ export class KeyRingService {
         mode: "direct",
         data: {
           ...data,
-          estimatedGasPrice: (data as any)?.gasPrice || estimatedGasPrice,
-          estimatedGasLimit: (data as any)?.gas || estimatedGasLimit,
-          decimals,
+          // estimatedGasPrice: (data as any)?.gasPrice ,
```
Comment (toandq2009) [inline packages/background/src/keyring/service.ts:820]: need remove this comment

File: packages/background/src/ledger/ledger.ts [hunk 8]
```diff
@@ -24,6 +24,8 @@ export const ledgerProxy = async (
         ledger = await LedgerInternal.init(currentMode, initArgs, ledgerType);
         response = true;
       } catch (error) {
+        console.log("error ledgerProxy", error);
```
Comment (toandq2009) [inline packages/background/src/ledger/ledger.ts:27]: need remove this console.log

File: packages/background/src/ledger/service.ts [hunk 9]
```diff
@@ -159,7 +159,7 @@ export class LedgerService {
           retryCount,
         };
       } catch (e) {
-        console.log(e);
+        console.log("initLedger e", e);
```
Comment (toandq2009) [inline packages/background/src/ledger/service.ts:162]: need remove this console or convert to console.error

File: packages/background/src/ledger/service.ts [hunk 10]
```diff
@@ -223,6 +223,8 @@ export class LedgerService {
 
           promises.push(aborter.wait());
 
+          console.log("initLedger pro 2", promises);
```
Comment (toandq2009) [inline packages/background/src/ledger/service.ts]: need remove this console

File: packages/background/src/updater/service.ts [hunk 11]
```diff
@@ -160,157 +160,45 @@ export class ChainUpdaterService {
...
       result: {
         node_info: {
-          network: "osmosis-1";
+          network: string;
         };
       };
     }>("/status");
 
-    const resultChainId = result.data.result.node_info.network;
+    const resultChainId = result?.data?.result?.node_info?.network;
```
Comment (toandq2009) [inline packages/background/src/updater/service.ts]: Using get function from lodash

File: packages/background/src/updater/service.ts [hunk 12]
```diff
@@ -160,157 +160,45 @@ export class ChainUpdaterService {
...
 
     const version = ChainIdHelper.parse(chainId);
     const fetchedVersion = ChainIdHelper.parse(resultChainId);
 
     // TODO: Should throw an error?
-    if (version.identifier !== fetchedVersion.identifier) {
+    if (version?.identifier !== fetchedVersion.identifier) {
```
Comment (toandq2009) [inline packages/background/src/updater/service.ts:181]: using guard code or get method from lodash

**Review by toandq2009 — COMMENTED** [review]
need check a some problem

**Review by toandq2009 — APPROVED** [review]
lgtm

**Pull request merged**